File: java/src/tarehart/rlbot/steps/debug/TagAlongStep.java
package tarehart.rlbot.steps.debug;

import tarehart.rlbot.AgentInput;
import tarehart.rlbot.AgentOutput;
import tarehart.rlbot.input.CarData;
import tarehart.rlbot.math.vector.Vector2;
import tarehart.rlbot.planning.Plan;
import tarehart.rlbot.routing.PositionFacing;
import tarehart.rlbot.steps.BlindStep;
import tarehart.rlbot.steps.Step;
import tarehart.rlbot.steps.travel.SlideToPositionStep;
import tarehart.rlbot.time.Duration;

import java.awt.*;
import java.util.Optional;

public class TagAlongStep implements Step {
    private Plan plan;

    public Optional<AgentOutput> getOutput(AgentInput input) {

        if (plan != null && !plan.isComplete()) {
            Optional<AgentOutput> output = plan.getOutput(input);
            if (output.isPresent()) {
                return output;
            }
        }

        plan = new Plan().withStep(new SlideToPositionStep(in -> {
            Optional<CarData> enemyCarOption = in.getEnemyCarData();
            CarData enemyCar = enemyCarOption.get();

            Vector2 waypoint = enemyCar.getPosition().plus(enemyCar.getOrientation().getRightVector().scaled(4)).flatten();
            Vector2 targetFacing = enemyCar.getOrientation().getNoseVector().flatten();
            return new PositionFacing(waypoint, targetFacing);
        })).withStep(new BlindStep(new AgentOutput().withAcceleration(1), Duration.ofSeconds(3)));

        return plan.getOutput(input);

    }

    @Override
    public boolean canInterrupt() {
        return false;
    }

    @Override
    public String getSituation() {
        return Plan.concatSituation("Tagging along", plan);
    }

    @Override
    public void drawDebugInfo(Graphics2D graphics) {
        if (Plan.activePlan(plan).isPresent()) {
            plan.getCurrentStep().drawDebugInfo(graphics);
        }
    }
}

File: src/main/java/tarehart/rlbot/steps/debug/TagAlongStep.kt
package tarehart.rlbot.steps.debug

import tarehart.rlbot.AgentOutput
import tarehart.rlbot.TacticalBundle
import tarehart.rlbot.planning.Plan
import tarehart.rlbot.tactics.TacticsTelemetry
import tarehart.rlbot.routing.PositionFacing
import tarehart.rlbot.steps.blind.BlindStep
import tarehart.rlbot.steps.NestedPlanStep
import tarehart.rlbot.steps.travel.ParkTheCarStep
import tarehart.rlbot.time.Duration

class TagAlongStep : NestedPlanStep() {

    override fun doComputationInLieuOfPlan(bundle: TacticalBundle): AgentOutput? {
        val p = Plan().withStep(ParkTheCarStep { inp ->
            val enemyCar = TacticsTelemetry[inp.playerIndex]?.enemyPlayerWithInitiative?.car!!
            val waypoint = enemyCar.position.plus(enemyCar.orientation.rightVector.scaled(4.0)).flatten()
            val targetFacing = enemyCar.orientation.noseVector.flatten()
            PositionFacing(waypoint, targetFacing)
        }).withStep(BlindStep(Duration.ofSeconds(0.2), AgentOutput()))

        return startPlan(p, bundle)
    }

    override fun getLocalSituation(): String {
        return "Tagging along"
    }
}
